### PR TITLE
[python][scikit-learn] Support for multiple evaluation metrics

### PR DESCRIPTION
**Pull request by gramirezespinoza**
This PR:
* **train, cv**: Allows for multiple custom evaluation metrics
* **scikit-learn API**: Allows multiple custom evaluation metrics
* **scikit-learn API**: Allows using a mix of custom evaluation functions and _LGBM's evaluation metrics_ as defined [here](https://lightgbm.readthedocs.io/en/latest/Parameters.html#metric-parameters)
* **scikit-learn API**: Fixes a bug that prevented the use of a list of _LGBM's evaluation metrics_ in LGBMClassifier

The `training` and `cv` APIs allow the use of multiple _LGBM's evaluation metrics_ defined [here](https://lightgbm.readthedocs.io/en/latest/Parameters.html#metric-parameters) but allow only one custom evaluation function. With this changes the user will be able to monitor multiple custom evaluation functions.

A similar situation happens in the scikit-learn API. The `eval_metric` parameter in the `fit` function accepts a list of _LGBM's evaluation metrics_ or one callable. With these changes the user will be able to monitor multiple _LGBM's evaluation metrics_, custom metrics, or a mix of both.

### CODE REVIEW

**Comment by guolinke** [comment]
@gramirezespinoza can you merge the latest master branch to pass CI?
@StrikerRUS what is your opinion for this PR?

**Comment by StrikerRUS** [comment]
> @StrikerRUS what is your opinion for this PR?

@gramirezespinoza  Can you please clarify the intent of this PR? Is it a refactoring? Because all points listed in the starting comment for this PR, like multiple custom metrics and the mix of custom and built-in metrics, are already supported in LightGBM.

**Comment by gramirezespinoza** [comment]
hey @StrikerRUS, @guolinke thanks for having a look at the PR!

The idea of the PR is to have total flexibility on the number and types of metrics that the user can monitor during training. To the best of my knowledge, the functionality on this PR is not yet implemented.

Happy to answer more questions if needed.

**All details**
Will start with the most important: the PR fixes a small bug in the `LGBMClassifier` class. The code as-is breaks when `eval_metric` is a list of strings. This is meant to work (as referenced in the documentation) and does work on other derived classes of `LGBMModel` except LGBMClassifier. The issue is in the following code in the master branch:

https://github.com/microsoft/LightGBM/blob/1e2013a306c0ca719fa076a733796a6372e5b77e/python-package/lightgbm/sklearn.py#L785-L798

with the fix the following line works as expected (see tests in PR):

`gbm = lgb.LGBMClassifier(**params).fit(eval_metric=['fair', 'error'], **params_fit)`


The second, is added functionality to allow mix of "string" metrics and custom metrics in the sklearn API. Now, it is possible to do this (see tests in PR):

`gbm = lgb.LGBMClassifier(**params).fit(eval_metric=[custom_recall, custom_precision, "fair"], **params_fit)`

where `custom_recall`, and `custom_precision` are callables.

Finally, the `train` and `cv` python APIs are modified to allow multiple `feval` like this:

```
model = lgb.train(
    ...
    feval=[custom_recall, custom_precision],
    ...)
```

For any other examples, please see the tests in the PR.

**Comment by gramirezespinoza** [comment]
> @gramirezespinoza can you merge the latest master branch to pass CI?

done

**Comment by StrikerRUS** [comment]
@gramirezespinoza OK, seems I got it! This PR is a bug fix for 

> Will start with the most important: the PR fixes a small bug in the LGBMClassifier class. The code as-is breaks when eval_metric is a list of strings.

and a refactoring for better usage of mix of custom and built-in metrics in one parameter (namely, `feval`/`eval_metric`), because right now one should use different arguments to achieve the goal: https://github.com/microsoft/LightGBM/issues/2182#issuecomment-493452818.

Will it make sense to split this PR into two?

**Comment by gramirezespinoza** [comment]
@StrikerRUS sure will split this into two PRs

**Comment by gramirezespinoza** [comment]
Bugfix of LGBMClassifier in [this PR](https://github.com/microsoft/LightGBM/pull/3222)

**Comment by gramirezespinoza** [comment]
Second part of this PR implemented in [3254](https://github.com/microsoft/LightGBM/pull/3254).

Closing this PR.
